chore: bundle vendors in internal packages

millsp opened the pull request:
Replaces https://github.com/prisma/prisma/pull/20857

Part of https://github.com/prisma/team-orm/issues/106

This PR aims to solve the fact that we are not able to update to newer ESM dependencies. To achieve this, we are switching to a bundling-only strategy (chunked). By bundling both CJS and ESM dependencies together with our published code, we can eliminate the semantic incompatibilities between both and produce a bundle that is CJS only (and possibly ESM bundles in the future, too).

In contrast with the previous PR, this one does not attempt to bundle into a single file. Instead it uses what is called bundle chunking, called `splitting` in `esbuild`. This basically allows a set of file to have their dependencies bundled (what we want to solve the ESM issue), and the bundle is efficiently shared between all existing files (chunked), and all files are efficiently chunked themselves. This fully preserves the multi-file structure that we have today, so we won't impact dependents with this change.

This PR is substantially simpler than the original one. 

/integration

millsp committed:
chore: bundle vendors

millsp committed:
update locks

millsp committed:
fix tsc plugin to only include dev into bundle

millsp committed:
fix local postinstall

millsp committed:
contribute to smaller cli bundles

millsp committed:
migrate: make prompts lib external

millsp committed:
leave generator helper to be fully bundled

millsp committed:
update locks

millsp committed:
mark prompts external in internals too

millsp committed:
mark arg external in internals

millsp committed:
don't bundle debug for generator-helper

millsp committed:
simplify e2e run.ts to include all

millsp committed:
update lockfile

millsp committed:
fix e2e

millsp committed:
add missing tmp dest

millsp committed:
fix paths in e2e

millsp committed:
fix missing slash in name

millsp committed:
unskip tests

Jolg42 commented on packages/client/tests/e2e/_utils/run.ts:
```diff
@@ -135,12 +109,8 @@ async function main() {
-    `/tmp/prisma-client-0.0.0.tgz:/tmp/prisma-client-0.0.0.tgz`,
-    `/tmp/prisma-engines-0.0.0.tgz:/tmp/prisma-engines-0.0.0.tgz`,
-    `/tmp/prisma-generator-helper-0.0.0.tgz:/tmp/prisma-generator-helper-0.0.0.tgz`,
-    `/tmp/prisma-nextjs-monorepo-workaround-plugin-0.0.0.tgz:/tmp/prisma-nextjs-monorepo-workaround-plugin-0.0.0.tgz`,
```
That's a really nice cleanup 💚 

It would help for https://github.com/prisma/team-orm/issues/595

Jolg42 commented on .gitignore:
```diff
@@ -40,3 +40,5 @@ junit.xml
 
 .idea/
+
+copy.package.json
```
nit: it's usually better to add a suffix to the file name so the files end up next to each other in local dev
```suggestion
package.copy.json
```
or so.

millsp committed:
mark packages external

millsp committed:
Merge branch 'main' of github.com:prisma/prisma into chore/bundle-vendors

millsp committed:
apply review comments

millsp committed:
make engines-version to be bundled in client

millsp committed:
update lock

Jolg42 commented:
Oups, merge conflict now.
I'll review this tomorrow.

I don't understand why the integration release didn't trigger here
```
 Pull Request Body contains /integration: false
```
https://github.com/prisma/prisma/actions/runs/6933779032/job/18862216538?pr=22017#step:5:11

So I started it manually here 
https://github.com/prisma/prisma/actions/workflows/release-ci.yml?query=branch%3Achore%2Fbundle-vendors
<img width="336" alt="Screenshot 2023-11-20 at 19 35 14" src="https://github.com/prisma/prisma/assets/1328733/1f8cf97f-27f4-4269-b446-73af4bf4bda2">

Jolg42 commented:
@millsp It looks like it breaks the publish script, could you have a look?
```
Error: TypeError: Cannot read properties of undefined (reading 'match')
    at getEnginesCommitHash (/home/runner/work/prisma/prisma/scripts/ci/publish.ts:660:40)
    at publish (/home/runner/work/prisma/prisma/scripts/ci/publish.ts:615:33)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
 ELIFECYCLE  Command failed with exit code 1.
```
https://github.com/prisma/prisma/actions/runs/6934381508/job/18862354471#step:5:2198

Jolg42 commented:
@millsp Could you check all the packages sizes before and after?

Quick check for some for me Prisma Client didn't change.

CLI got a bit smaller:
<img width="304" alt="Screenshot 2023-11-20 at 19 51 02" src="https://github.com/prisma/prisma/assets/1328733/d267c9d2-ff1e-4622-a798-26404d289b4c">
<img width="299" alt="Screenshot 2023-11-20 at 19 50 21" src="https://github.com/prisma/prisma/assets/1328733/9f7404a2-3184-4075-8e36-33d7112d37d5">

`@prisma/migrate` got bigger:
<img width="305" alt="Screenshot 2023-11-20 at 19 52 15" src="https://github.com/prisma/prisma/assets/1328733/66ac599a-a39a-4319-b4a4-6f552c6bf3f8">
<img width="317" alt="Screenshot 2023-11-20 at 19 52 02" src="https://github.com/prisma/prisma/assets/1328733/788ffdd6-8330-4120-babd-c0c9196d2b52">

....

millsp committed:
fix publish script

millsp committed:
Merge branch 'main' of github.com:prisma/prisma into chore/bundle-vendors

millsp commented:
Thanks for catching that @Jolg42, I fixed the publish script. There will be 1 test that will fail in ecosystem tests, because we try to use `@prisma/engine-version` in `@prisma/client`, I have changed things there. Before this PR `@prisma/engine-version` was both bundled if `@prisma/client` and a dependency of it, that does not make a ton of sense to me as only the bundled one will be used. I will thus adapt the ecosystem test (probably a 1-line change).

millsp commented:
As you suggested, I did a before and after

```
prisma-debug-5.7.0-dev.31.tgz                             7.2 kB  
prisma-driver-adapter-utils-5.7.0-dev.31.tgz              7.8 kB  
prisma-nextjs-monorepo-workaround-plugin-5.7.0-dev.31.tgz 6.2 kB  
prisma-generator-helper-5.7.0-dev.31.tgz                  11.3 kB 
prisma-get-platform-5.7.0-dev.31.tgz                      16.3 kB 
prisma-adapter-libsql-5.7.0-dev.31.tgz                    8.8 kB  
prisma-adapter-neon-5.7.0-dev.31.tgz                      10.3 kB 
prisma-adapter-pg-5.7.0-dev.31.tgz                        9.3 kB  
prisma-adapter-planetscale-5.7.0-dev.31.tgz               9.0 kB  
prisma-fetch-engine-5.7.0-dev.31.tgz                      14.9 kB 
prisma-engines-5.7.0-dev.31.tgz                           872.8 kB
prisma-internals-5.7.0-dev.31.tgz                         77.7 kB 
prisma-instrumentation-5.7.0-dev.31.tgz                   7.9 kB  
prisma-migrate-5.7.0-dev.31.tgz                           65.6 kB 
prisma-client-5.7.0-dev.31.tgz                            1.3 MB  
prisma-5.7.0-dev.31.tgz                                   5.1 MB  
```

```
prisma-debug-5.7.0-integration-chore-bundle-vendors.2.tgz                              13.1 kB 
prisma-driver-adapter-utils-5.7.0-integration-chore-bundle-vendors.2.tgz               7.9 kB  
prisma-nextjs-monorepo-workaround-plugin-5.7.0-integration-chore-bundle-vendors.2.tgz  6.2 kB  
prisma-generator-helper-5.7.0-integration-chore-bundle-vendors.2.tgz                   16.4 kB 
prisma-get-platform-5.7.0-integration-chore-bundle-vendors.2.tgz                       118.2 kB
prisma-adapter-libsql-5.7.0-integration-chore-bundle-vendors.2.tgz                     8.8 kB  
prisma-adapter-neon-5.7.0-integration-chore-bundle-vendors.2.tgz                       10.3 kB 
prisma-adapter-pg-5.7.0-integration-chore-bundle-vendors.2.tgz                         9.4 kB  
prisma-adapter-planetscale-5.7.0-integration-chore-bundle-vendors.2.tgz                9.0 kB  
prisma-fetch-engine-5.7.0-integration-chore-bundle-vendors.2.tgz                       428.4 kB
prisma-engines-5.7.0-integration-chore-bundle-vendors.2.tgz                            22.5 kB 
prisma-internals-5.7.0-integration-chore-bundle-vendors.2.tgz                          726.9 kB
prisma-instrumentation-5.7.0-integration-chore-bundle-vendors.2.tgz                    8.9 kB  
prisma-migrate-5.7.0-integration-chore-bundle-vendors.2.tgz                            1.3 MB  
prisma-client-5.7.0-integration-chore-bundle-vendors.2.tgz                             1.3 MB  
prisma-5.7.0-integration-chore-bundle-vendors.2.tgz                                    5.4 MB  
```

millsp committed:
ensure proper tree-shaking

millsp committed:
update snap

millsp commented on helpers/compile/plugins/resolvePathsPlugin.ts:
```diff
@@ -0,0 +1,59 @@
+  return []
+}
+
+/**
+ * Forces `esbuild` to always use the TS compiler paths, even when we are
+ * bundling a local dependency of a local dependency, ensuring maximum
+ * tree-shaking. Note: `esbuild` has some support for this, though it is limited
+ * in the amount of dependency nesting it supports.
+ */
```
This helped shrinking the CLI by nearly 1MB (excluding the minification). The minification + this change cut the CLI size by nearly half.

millsp committed:
respect external packages in path resolution

millsp commented on packages/client/package.json:
```diff
@@ -175,8 +176,5 @@
       "optional": true
     }
   },
-  "dependencies": {
-    "@prisma/engines-version": "5.7.0-10.7025c28a99be33ac6e150c344958fbcd0b7ffa9c"
-  },
```
See the comment about this in the review directly.

millsp replied:
https://github.com/prisma/prisma/pull/22017#issuecomment-1819688351

millsp committed:
simplify and document changes to tscPlugin

millsp commented:
It is important to note that this is one major element to unblock ESM updates, yet it won't do everything. For instance, `ts-node` is incompatible with ESM down the line, or the dependency had breaking changes other than just ESM, etc... This should unblock the "easy" cases and it will be interesting to see where/if things block next.

Jolg42 reviewed (approved):
🙌🏼 :shipit:
